### ISSUE TITLE
new Check: RegExpComment

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#8137]
summary from #7388

Although spaces after slashes are optional in the comments, their absence makes the documentation difficult to read. It is the de facto standard to put at least one whitespace after the leading two slashes like so:

`// comment`

and it is style violation to omit the space like so:

`//comment`

This convention is implied in Section 3.7 "Comments" of the [Java Language Specification](https://docs.oracle.com/javase/specs/jls/se13/html/jls-3.html#jls-3.7):

>    3.7. Comments
> 
> There are two kinds of comments:
> 
>     /* text */
> 
>     A traditional comment: all the text from the ASCII characters /* to the ASCII characters */ is ignored (as in C and C++).
> 
>     // text
> 
>     An end-of-line comment: all the text from the ASCII characters // to the end of the line is ignored (as in C++).

It appears to be *de rigueur* in Java itself, for example in http://hg.openjdk.java.net/jdk7u/jdk7u6/jdk/file/8c2c5d63a17e/src/share/classes/java/lang/String.java 

```
/var/tmp $ javac TestClass.java

public class TestClass {
    //no whitespace after double slash here.
}
```

**Expected violation:** TestClass.java:2:5 Missing a whitespace after the leading double slash.

Check will NOT run on javadocs 

So imaginary beharior could be like:
```
$ cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
                        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="RegExpComment_nonexisting_check">
      <property name="format" value="^([^\s\/*])"/>
     <property name="ignoreCase" value="false">
     <property name="ignorePattern" value="(^/*$)|(^* )">
      <message key="todo.match" value="Comment should start with space."/>
    </module>
  </module>
</module>

$ cat Test.java 
////////////////////////////////////////
// above line is ok
public class Test {
  /** ok */
  public Test() {} //violation
  /**violation */
  public Test(int a) {;} // ok
  public Test(char a) {;} /*violation*/
  public Test(String a) {;} /* violation*/

}

$ java -jar /var/tmp/checkstyle-8.29-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /var/tmp/Test.java:2: Comment should start with space. [RegExpComment]
[ERROR] /var/tmp/Test.java:4: Comment should start with space. [RegExpComment]
Audit done.
Checkstyle ends with 2 errors.

```

**Comment by Gaurav-Punjabi** [comment]
I'm on it, once it's approved.

**Comment by koppor** [comment]
Thank you for working on it.

The `module name` reads strange. Never saw underscores in that. - I wonder why I have to config a regular expression. Why can't I just use a module like `SpaceAfterCommentStart`? (Even though it is a very specialized module)

Just for the record - the difference to following (currently working!) check:

```xml
<module name="TodoComment">
    <property name="id" value="commentStartWithSpace"/>
    <property name="format" value="^([^\s\/*])"/>
    <message key="todo.match" value="Comment text should start with space."/>
</module>
```

is

> Unfortunately there are bunch of wontfixes for cases like /////// or /** so regexp above is not very reliable, but it should not catch such cases.


([Source](https://github.com/checkstyle/checkstyle/issues/7388#issuecomment-581203062))

I have issues with code like follows

```java
/*
public class CausingViolationBecauseLineStartsWithCharacter {
}
*/
```

Would that be solved when using `RegExpComment_nonexisting_check` instead of `TodoComment`?

**Comment by koppor** [comment]
Note that the above rule was applied at https://github.com/checkstyle/checkstyle/commit/dbac7fb8e605c3f29172d7f4a894b1b91e380edc. One could use the parent commit as test case maybe?

**Comment by romani** [comment]
> Why can't I just use a module like SpaceAfterCommentStart?

we thy to make module more general to be reused to catch other not good comments.

> `<module name="RegExpComment_nonexisting_check">`

I did bad name on purpose to not anybody think it is already implemented. Sorry for confusion.

> Would that be solved when using RegExpComment_nonexisting_check instead of TodoComment?

we can make Check to have tokens configured to let user run it only on `//` (single line comment), or or only on multiline comment `/* */`, or both.

**Comment by romani** [comment]
@pbludov , @strkkk , can you read this proposal and share ideas to let us approve issue.

**Comment by strkkk** [comment]
@romani it is not clear is this check only for single line comments or not? if yes, it is not clear what should be a violation. Also, description is a bit misleading since it says that check should skip javadocs (why?) and in code example you have javadoc with violation comment.

**Comment by romani** [comment]
all javadoc comments are block comments:
```
$ java -jar checkstyle-8.36.2-all.jar -T  Test.java
CLASS_DEF -> CLASS_DEF [3:0]
|--MODIFIERS -> MODIFIERS [3:0]
|   |--SINGLE_LINE_COMMENT -> // [1:0]
|   |   `--COMMENT_CONTENT -> //////////////////////////////////////\n [1:2]
|   |--SINGLE_LINE_COMMENT -> // [2:0]
|   |   `--COMMENT_CONTENT ->  above line is ok\n [2:2]
|   `--LITERAL_PUBLIC -> public [3:0]
|--LITERAL_CLASS -> class [3:7]
|--IDENT -> Test [3:13]
`--OBJBLOCK -> OBJBLOCK [3:18]
    |--LCURLY -> { [3:18]
    |--CTOR_DEF -> CTOR_DEF [5:2]
    |   |--MODIFIERS -> MODIFIERS [5:2]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [4:2]
    |   |   |   |--COMMENT_CONTENT -> * ok  [4:4]
    |   |   |   `--BLOCK_COMMENT_END -> */ [4:8]
    |   |   `--LITERAL_PUBLIC -> public [5:2]
    |   |--IDENT -> Test [5:9]
    |   |--LPAREN -> ( [5:13]
    |   |--PARAMETERS -> PARAMETERS [5:14]
    |   |--RPAREN -> ) [5:14]
    |   `--SLIST -> { [5:16]
    |       `--RCURLY -> } [5:17]
    |--CTOR_DEF -> CTOR_DEF [7:2]
    |   |--MODIFIERS -> MODIFIERS [7:2]
    |   |   |--SINGLE_LINE_COMMENT -> // [5:19]
    |   |   |   `--COMMENT_CONTENT -> violation\n [5:21]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [6:2]
    |   |   |   |--COMMENT_CONTENT -> *violation  [6:4]
    |   |   |   `--BLOCK_COMMENT_END -> */ [6:14]
    |   |   `--LITERAL_PUBLIC -> public [7:2]
    |   |--IDENT -> Test [7:9]
    |   |--LPAREN -> ( [7:13]
    |   |--PARAMETERS -> PARAMETERS [7:14]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [7:14]
    |   |       |--MODIFIERS -> MODIFIERS [7:14]
    |   |       |--TYPE -> TYPE [7:14]
    |   |       |   `--LITERAL_INT -> int [7:14]
    |   |       `--IDENT -> a [7:18]
    |   |--RPAREN -> ) [7:19]
    |   `--SLIST -> { [7:21]
    |       |--EMPTY_STAT -> ; [7:22]
    |       `--RCURLY -> } [7:23]
    |--CTOR_DEF -> CTOR_DEF [8:2]
    |   |--MODIFIERS -> MODIFIERS [8:2]
    |   |   |--SINGLE_LINE_COMMENT -> // [7:25]
    |   |   |   `--COMMENT_CONTENT ->  ok\n [7:27]
    |   |   `--LITERAL_PUBLIC -> public [8:2]
    |   |--IDENT -> Test [8:9]
    |   |--LPAREN -> ( [8:13]
    |   |--PARAMETERS -> PARAMETERS [8:14]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [8:14]
    |   |       |--MODIFIERS -> MODIFIERS [8:14]
    |   |       |--TYPE -> TYPE [8:14]
    |   |       |   `--LITERAL_CHAR -> char [8:14]
    |   |       `--IDENT -> a [8:19]
    |   |--RPAREN -> ) [8:20]
    |   `--SLIST -> { [8:22]
    |       |--EMPTY_STAT -> ; [8:23]
    |       `--RCURLY -> } [8:24]
    |--CTOR_DEF -> CTOR_DEF [9:2]
    |   |--MODIFIERS -> MODIFIERS [9:2]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [8:26]
    |   |   |   |--COMMENT_CONTENT -> violation [8:28]
    |   |   |   `--BLOCK_COMMENT_END -> */ [8:36]
    |   |   `--LITERAL_PUBLIC -> public [9:2]
    |   |--IDENT -> Test [9:9]
    |   |--LPAREN -> ( [9:13]
    |   |--PARAMETERS -> PARAMETERS [9:14]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [9:14]
    |   |       |--MODIFIERS -> MODIFIERS [9:14]
    |   |       |--TYPE -> TYPE [9:14]
    |   |       |   `--IDENT -> String [9:14]
    |   |       `--IDENT -> a [9:21]
    |   |--RPAREN -> ) [9:22]
    |   `--SLIST -> { [9:24]
    |       |--EMPTY_STAT -> ; [9:25]
    |       `--RCURLY -> } [9:26]
    |--BLOCK_COMMENT_BEGIN -> /* [9:28]
    |   |--COMMENT_CONTENT ->  violation [9:30]
    |   `--BLOCK_COMMENT_END -> */ [9:39]
    `--RCURLY -> } [11:0]
```

we can think of some extra simple code to skip comments with javadoc identifier in block comments.

**Comment by pbludov** [comment]
For a block comment, there must be a space before the closing`*/'. But single-line comments will definitely not have it. This means that we need one regular expression for single-line comments and another for block comments. And it will be easier to create two new checks.

**Comment by romani** [comment]
@AkMo3 , is this issue covered in most cases by #10107 ?

**Comment by AkMo3** [comment]
@romani Yes. 

> there must be a space before the closing`*/

This can be achieved by something like `//BLOCK_COMMENT_BEGIN[./COMMENT_CONTENT[not(ends-with (@text, ' *'))]]`
By this, we will be getting all the block comments that do not end with a ` *`(space before the asterisk)

> Space before Single Line comment and braces inside code

These cases had been achieved in #10107
